plugins/filter_aws/aws.c

```diff
@@ -698,23 +751,180 @@ static int get_ec2_metadata(struct flb_filter_aws *ctx)
+    ret = fetch_func(ctx);
+    if (ret == 0) {
+        group->done = FLB_TRUE;
+    }
+    return ret;
+}
+
+/*
+ * Fetches all metadata values, including tags, from IMDS.
+ * Function handles retries as configured for each metadata group.
+ *
+ * Returns 0 on success, negative values on failures.
+ * Returns FLB_FILTER_AWS_CONFIGURATION_ERROR in case of configuration error.
+ */
+static int get_ec2_metadata(struct flb_filter_aws *ctx)
+{
+    int ret;
+    int metadata_fetched = FLB_TRUE;
+
+    if (ctx->metadata_retrieved) {
+        return 0;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_instance_id,
+                                 get_ec2_metadata_instance_id);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_account_id,
+                                 get_ec2_metadata_account_id);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
         }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_instance_type,
+                                 get_ec2_metadata_instance_type);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_private_ip,
+                                 get_ec2_metadata_private_ip);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_vpc_id, get_ec2_metadata_vpc_id);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_ami_id, get_ec2_metadata_ami_id);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_hostname, get_ec2_metadata_hostname);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_az, get_ec2_metadata_az);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    ret = get_ec2_metadata_group(ctx, &ctx->group_tag, get_ec2_metadata_tags);
+    if (ret < 0) {
+        if (ret == FLB_FILTER_AWS_CONFIGURATION_ERROR) {
+            return ret;
+        }
+        metadata_fetched = FLB_FALSE;
+    }
+
+    if (metadata_fetched) {
```

Comment:
nit: Eduardo says to always use `if (metadata_fetched == FLB_TRUE)` as FLB code style

Reply:
https://github.com/fluent/fluent-bit/pull/7245/commits/9216df6d92669686756f513a05291d1a29108f7f